internal/receiver/discoveryreceiver/statement_evaluator.go

```diff
@@ -0,0 +1,220 @@
+				return map[string]any{"msg": pattern}
+			},
+		},
+		encoder: encoder,
+	}
+	var err error
+	if le.evaluatedLogger, err = zapConfig.Build(
+		// zap.OnFatal must not be WriteThenFatal or WriteThenNoop since it's rewritten to be WriteThenFatal
+		// https://github.com/uber-go/zap/blob/e06e09a6d396031c89b87383eef3cad6f647cf2c/logger.go#L315.
+		// Using an arbitrary action offset.
+		zap.WithFatalHook(zapcore.WriteThenFatal+100),
+		zap.WrapCore(func(core zapcore.Core) zapcore.Core { return le }),
+	); err != nil {
+		return nil, err
+	}
+	return le, nil
+}
+
+// Enabled is a zapcore.Core method. We should be enabled for all
+// levels since we want to intercept all statements.
+func (se *statementEvaluator) Enabled(zapcore.Level) bool {
+	return true
+}
+
+// With is a zapcore.Core method. We clone ourselves so all
+// modified downstream loggers are still evaluated.
+func (se *statementEvaluator) With(fields []zapcore.Field) zapcore.Core {
+	cp := *se
+	cp.encoder = se.encoder.Clone()
+	for i := range fields {
+		fields[i].AddTo(cp.encoder)
+	}
+	return &cp
+}
+
+// Check is a zapcore.Core method. Similar to Enabled() we want to
+// return a valid CheckedEntry for all logging attempts to intercept
+// all statements.
+func (se *statementEvaluator) Check(entry zapcore.Entry, checkedEntry *zapcore.CheckedEntry) *zapcore.CheckedEntry {
+	return checkedEntry.AddCore(entry, se)
+}
+
+// Write is a zapcore.Core method. This is where the logged entry
+// is converted to a statussources.Statement, if from a downstream receiver,
+// and its content is evaluated for status matches and plog.Logs translation/submission.
+func (se *statementEvaluator) Write(entry zapcore.Entry, fields []zapcore.Field) error {
+	statement, err := statussources.StatementFromZapCoreEntry(se.encoder, entry, fields)
+	if err != nil {
+		return err
+	}
+	if name, ok := statement.Fields["name"]; ok {
+		if cid, err := config.NewComponentIDFromString(fmt.Sprintf("%v", name)); err == nil {
+			if cid.Type() == "receiver_creator" && cid.Name() == se.config.ID().String() {
+				// this is from our internal Receiver Creator and not a generated receiver, so write
+				// it to our logger core without submitting the entry for evaluation
+				if ce := se.logger.Check(entry.Level, ""); ce != nil {
+					// forward to our logger now that we know entry.Level is accepted
+					se.logger.Core().Write(entry, fields)
+				}
+				return nil
+			}
+		}
+	}
+
+	if pLogs := se.evaluateStatement(statement); pLogs.LogRecordCount() > 0 {
+		se.pLogs <- pLogs
+	}
+
+	return nil
+}
+
+// Sync is a zapcore.Core method.
+func (se *statementEvaluator) Sync() error {
+	return nil
+}
+
+// evaluateStatement will convert the provided statussources.Statement into a plog.LogRecord
+// and match it against the applicable configured ReceiverEntry's status Statement.[]Match
+func (se *statementEvaluator) evaluateStatement(statement *statussources.Statement) plog.Logs {
```

Comment:
Would it be better to break this method up into a few functions or methods?

Reply:
will take a crack at this with a plog preparation helper.

Reply:
updated